algorithms/scripts/BernoulliRBM-LogisticRegressionCV-yl565.py

```diff
@@ -0,0 +1,199 @@
+X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=0)
+'Size: {:,} features, {:,} training samples, {:,} testing samples'.format(len(X.columns), len(X_train), len(X_test))
+
+
+# ## Class and functions for classification and evaluation
+
+# In[3]:
+
+class NormDist(BaseEstimator, TransformerMixin):
+    """Scale data into [0, 1] range
+    Fit each column of X with a Normal distribution and transform the elements to
+    probability values according to the fitted cdf
+    """
+    def __init__(self):
+        pass
+
+    def fit(self, X, y=None):
+        X = X.astype(float)
+        self.params = []
+        for col in X.T:
+            self.params.append(norm.fit(col))
+
+    def transform(self, X, y=None):
+        X = X.astype(float)
+        for i, col in enumerate(X.T):
+            X[:, i] = norm.cdf(col, *self.params[i])
+        return X    
+
+    def fit_transform(self, X, y=None):
+        self.fit(X, y)
+        return self.transform(X)
+
+    
+def fs_mad(x, y):
+    """    
+    Get the median absolute deviation (MAD) for each column of x
+    """
+    scores = mad(x) 
+    return scores, np.array([np.NaN]*len(scores))
+
+
+def gen_pipeline(n_feat, n_RBM_cmp):
+    """Generate classification pipeline
+    Steps:
+        1. Select n_feat number of features with highest MAD.
+        2. Scale the n_feat features to within 0-1 range.
+        3. Fitting a Restricted Boltzmann Machine (RBM) with n_feat visible units and n_RBM_cmp hidden units.
+        4. Train a logistic regression classifier with the output of RBM as input features.
+    """
+    # Step 1: feature selection
+    if n_feat is None:
+        pipe=[]  # Skip feature selection, use all features
+    else:
+        pipe=[SelectKBest(fs_mad, k=n_feat)]
+    
+    if n_RBM_cmp is not None:
+        # Step 2: feature scaling
+        pipe.append(NormDist())  # Mapping input feature to [0, 1] range assuming columns are Guassian
+
+        # Step 3: Feature learning
+        pipe.append(
+            BernoulliRBM(random_state=0,
+                         n_components=n_RBM_cmp,
+                         verbose=True,
+                         learning_rate=0.01,
+                         n_iter=30))
+    
+    # Step 4: Classification
+    pipe.append(LogisticRegressionCV(class_weight='balanced', random_state=0))
+    
+    return make_pipeline(*tuple(pipe))
+
+
+def eval_model(fitted_model, X_train, X_test, y_train, y_test, title_str):
+    """Evaluate and plot the ROC curve of fitted_model
+    """
+    y_pred_train = fitted_model.decision_function(X_train)
+    y_pred_test = fitted_model.decision_function(X_test)
+
+    def get_threshold_metrics(y_true, y_pred):
+        roc_columns = ['fpr', 'tpr', 'threshold']
+        roc_items = zip(roc_columns, roc_curve(y_true, y_pred))
+        roc_df = pd.DataFrame.from_items(roc_items)
+        auroc = roc_auc_score(y_true, y_pred)
```

Comment:
yeah, I have starred at that source code longer than I'd like to...still confusing.